feat(pulseaudio): add support for source device

vincentbernat opened the pull request:
## What type of PR is this? (check all applicable)

* [ ] Refactor
* [X] Feature
* [ ] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Description

Update pulseaudio module to accept a source instead of a sink.

This is quite similar to #2238 but I had the feeling there was too
much code duplication. At first, I did try to rely heavily on
templating to avoid that and keep the number of lines changed to a
minimum, but my C++ foo was not good enough (I can push what I have
currently if someone is interested, but it does not compile).

In this attempt, only a few functions are templated (not the whole
class) and a few helper functions are introduced to implement the
differences.

Instead of using switch/case to implement the differences, helper
functions could be specified as an abstract class and implemented by
different classes. I think this is not worth the effort.

```ini
[module/speaker]
type = internal/pulseaudio
format-volume = <label-volume>
format-muted = <label-muted>
label-volume = speaker: %percentage: 2%%
label-muted = muted speaker: %percentage: 2%%

[module/micro]
type = internal/pulseaudio
source = "@DEFAULT_SOURCE"
format-volume = <label-volume>
format-muted = <label-muted>
label-volume = micro: %percentage: 2%%
label-muted = muted micro: %percentage: 2%%
```

Fix #1096.

## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
* [x] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

vincentbernat committed:
feat(pulseaudio): add support for source device

Update pulseaudio module to accept a source instead of a sink.

This is quite similar to #2238 but I had the feeling there was too
much code duplication. At first, I did try to rely heavily on
templating to avoid that and keep the number of lines changed to a
minimum, but my C++ foo was not good enough (I can push what I have
currently if someone is interested, but it does not compile).

In this attempt, only a few functions are templated (not the whole
class) and a few helper functions are introduced to implement the
differences.

Instead of using switch/case to implement the differences, helper
functions could be specified as an abstract class and implemented by
different classes. I think this is not worth the effort.

```ini
[module/speaker]
type = internal/pulseaudio
format-volume = <label-volume>
format-muted = <label-muted>
label-volume = speaker: %percentage: 2%%
label-muted = muted speaker: %percentage: 2%%

[module/micro]
type = internal/pulseaudio
source = "@DEFAULT_SOURCE"
format-volume = <label-volume>
format-muted = <label-muted>
label-volume = micro: %percentage: 2%%
label-muted = muted micro: %percentage: 2%%
```

Fix #1096.

vincentbernat commented:
After thinking a bit, I think I can make something a bit more concise!

vincentbernat committed:
feat(pulseaudio): filter out monitors and auto_null

For source, we don't want to get something for monitors. For sink, we
exclude auto_null. This way, if you don't have an output or a
microphone, you don't get the icon.

vincentbernat commented:
I am unable to make inheritance work. I have errors around the use of `factory_util::unique<pulseaudio>` which seems to not play nice when the concrete class is not `pulseaudio` but a subclass. Something about `operator=` not being here. Also, it doesn't help much for the number of line of code. Again, I can push the code if you want to try to debug. Surely, I am a bit limited by my C++ knowledge.

I'll stop working on this until I get a feedback for a direction to follow. I am currently using the module as is.

Zopieux commented on src/modules/pulseaudio.cpp:
```diff
@@ -106,6 +120,17 @@ namespace modules {
+    auto s_name = m_pulseaudio->get_name();
+    if (s_name.empty() ||
+        s_name == "auto_null" ||
+        (s_name.size() >= 8 && s_name.compare(s_name.size() - 8, s_name.size(), ".monitor") == 0)) {
```
Consider adding `ends_with` in [`string_util`](https://github.com/polybar/polybar/blob/master/include/utils/string.hpp#L64) and use it here.

vincentbernat replied:
Thanks. I would like to get an ack about the big picture first. Once it's OK, I'll do what you suggest.

patrick96 commented on src/adapters/pulseaudio.cpp:
```diff
@@ -97,6 +106,87 @@ const string& pulseaudio::get_name() {
+    case devicetype::SOURCE:
+      return "source"s;
+  }
+  throw pulseaudio_error("unreachable");
```
Maybe rather have the throw statement in the `default` case and also mention the `m_device_type`.

patrick96 reviewed (commented):
The general direction is good.

Subclassing would probably make all the switch statements obsolete, but your way is fine as well.

>  have errors around the use of factory_util::unique<pulseaudio> which seems to not play nice when the concrete class is not pulseaudio but a subclass.

Since `make_unique` actually creates an instance of the class, the template parameter needs to be the concrete subclass.

xoores commented:
Hi,

I want to tackle pulseaudio source/sink thing for my use as well. I will probably go from scratch & add it from current HEAD since this has been stale since 2021 :slightly_smiling_face: 

I did share some fixes/enhancements for backlight module - would you be interested in this as well?

BR
X.

vincentbernat commented:
Yes, sorry, I didn't followup. I am using this code without any issue. It only crashes when Pulseaudio is restarted. As it works for me, I don't have the motivation to go the extra mile to finish the PR. Feel free to take over or start from scratch.

xoores commented:
Okay, I'll take a closer look at what has been done and adapt it to the current master HEAD so it can be merged easily :slightly_smiling_face: I've already shared some improvements for backlight module so I will do that for pulse as well.

I tend to play with pulse/pipewire a lot so restarts are happening quite regularly so this would be one thing to fix for me :grin:

maplepy commented:
Hey @xoores, is this still something you're working on or did you give up on the project?

xoores commented:
> Hey @xoores, is this still something you're working on or did you give up on the project?

Hi! Did not give up but also did not really want to waste much time writing a code that would not be accepted (I have another pull opened for backlight module). :sweat_smile: 

I will probably get to this as well but I have some work-related tasks that I needed to prioritize :slightly_smiling_face:

maplepy commented:
> did not really want to waste much time writing a code that would not be accepted

oh yeah makes total sense :)

> I will probably get to this as well but I have some work-related tasks that I needed to prioritize slightly_smiling_face

sure thing! take the time you need bro, there's no rush aha. enjoy yourself, that's the only thing that matters :heart:

vincentbernat commented:
I have updated this PR to work against master. No other change. I can resume work on this if nobody else has stepped up with a better approach. From my understanding, the current PR is mostly good.

vincentbernat commented:
As it was difficult to maintain out-of-tree and also because this module does not detect 100% of source/sink changes (I had to restart polybar from time to time), I have switched to an ipc module (and to pipewire as well):

```ini
[module/audio]
type = custom/ipc
hook-0 = cat $XDG_RUNTIME_DIR/i3/audio.txt 2> /dev/null
initial = 1
```

Excerpt of my [i3-companion](https://github.com/vincentbernat/i3wm-configuration/blob/master/bin/i3-companion):

```python
@on(StartEvent, TailEvent(command="pactl -f json subscribe"))
@polybar("audio")
async def audio_update(i3, event, line=""):
    """Update audio sink or source volume/status in Polybar."""
    if '"on":"sink"' not in line and '"on":"source"' not in line and line != "":
        return
    proc = await asyncio.create_subprocess_exec(
        "wpctl",
        "status",
        "-k",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.DEVNULL,
    )
    stdout, _ = await proc.communicate()
    if proc.returncode != 0:
        logger.info("cannot get wpctl status")
        return

    # Audio
    #  ├─ Devices:
    #  │      43. Navi 48 HDMI/DP Audio Controller    [alsa]
    #  │      44. HD Pro Webcam C920                  [alsa]
    #  │      45. Starship/Matisse HD Audio Controller [alsa]
    #  │
    #  ├─ Sinks:
    #  │  *   55. Starship/Matisse HD Audio Controller Digital Stereo (IEC958) [vol: 0.65]
    #  │
    #  ├─ Sources:
    #  │
    #  ├─ Filters:
    #  │
    #  └─ Streams:
    #         70. spotify
    #              75. output_FR       > ALCS1200A Digital:playback_FR	[active]
    #              89. output_FL       > ALCS1200A Digital:playback_FL	[active]
    Device = collections.namedtuple("Device", ["name", "volume", "muted"])
    devices = {"source": None, "sink": None}
    in_audio_section = False
    current_device = None
    lines = stdout.decode().split("\n")
    for line in lines:
        if line == "Audio":
            in_audio_section = True
        elif line == "" and in_audio_section:
            break
        elif line == " ├─ Sinks:" and in_audio_section:
            current_device = "sink"
        elif line == " ├─ Sources:" and in_audio_section:
            current_device = "source"
        elif line == " │":
            current_device = None
        elif line.startswith(" │  *") and current_device is not None:
            mo = re.match(
                r"^ │  \*\s+\d+\.\s+(?P<name>.*?)\s+\[vol: (?P<volume>[0-9.]+)(?P<muted> MUTED)?\]",
                line,
            )
            if mo:
                devices[current_device] = Device(
                    name=mo.group("name").replace("%", "%\u200b"),
                    volume=int(float(mo.group("volume")) * 100),
                    muted=bool(mo.group("muted")),
                )
    status = []
    for kind, icon in [("source", "microphone"), ("sink", "speaker")]:
        device = devices[kind]
        if device is None:
            continue
        wpdevice = f"@DEFAULT_AUDIO_{kind.upper()}@"
        actions = [
            f"%{{A1:wpctl set-mute {wpdevice} toggle:}}",
            "%{A3:pavucontrol:}",
        ]
        if device.muted:
            state = f"{icons[f"{icon}-muted"]}"
        else:
            state = f"{icons[icon]} {device.volume:>2}%"
            actions += [
                f"%{{A4:wpctl set-volume {wpdevice} 5%+:}}",
                f"%{{A5:wpctl set-volume {wpdevice} 5%-:}}",
            ]
        state += f"%{{T5}} {device.name}%{{T-}}"
        end = "%{A}" * len(actions)
        status.append(f"{"".join(actions)}{state}{end}")

    return " ".join(status)
```
